// @rx-angular/state/esm2022/lib/provide-rx-state-config.mjs
import { InjectionToken } from '@angular/core';
import { defaultAccumulator, } from '@rx-angular/state/selections';
import { queueScheduler } from 'rxjs';
/**
 * Injection token for the default accumulator function.
 *
 * @example
 * providers: [
 *  {
 *   provide: RX_ACCUMULATOR_FN,
 *   useValue: (state, slice) => ({ ...state, ...slice })
 *  }
 * ]
 */
export const RX_ACCUMULATOR_FN = new InjectionToken('RX_ACCUMULATOR_FN', {
    providedIn: 'root',
    factory: () => defaultAccumulator,
});
/**
 * Provider function to specify a custom `AccumulationFn` for `RxState` to use.
 * @param fn
 */
export function withAccumulatorFn(fn) {
    return {
        kind: 'Accumulator',
        providers: [{ provide: RX_ACCUMULATOR_FN, useValue: fn }],
    };
}
/**
 * Injection token for the default state scheduler
 *
 * @example
 * providers: [
 *  {
 *   provide: RX_STATE_SCHEDULER,
 *   useValue: asapScheduler
 *  }
 * ]
 */
export const RX_STATE_SCHEDULER = new InjectionToken('RX_STATE_SCHEDULER', {
    providedIn: 'root',
    factory: () => queueScheduler,
});
/**
 * Provider function to specify a scheduler for `RxState` to perform state updates & emit new values.
 * @param scheduler
 */
export function withScheduler(scheduler) {
    return {
        kind: 'Scheduler',
        providers: [{ provide: RX_STATE_SCHEDULER, useValue: scheduler }],
    };
}
/**
 * Provider function to specify synchronous (no) scheduling for `RxState`. The state computations
 * will be fully synchronous instead of using the default `queueScheduler`
 */
export function withSyncScheduler() {
    return {
        kind: 'Scheduler',
        providers: [{ provide: RX_STATE_SCHEDULER, useValue: 'sync' }],
    };
}
/**
 * This function is used to provide the configuration for the rxState function.
 *
 * You can provide multiple configurations at once.
 *
 * You can use these functions to provide the configuration:
 * - withAccumulatorFn - to provide a custom accumulator function
 * - withScheduler - to provide a custom scheduler
 *
 */
export function provideRxStateConfig(...configs) {
    return flatten(configs.map((c) => c.providers));
}
function flatten(arr) {
    return arr.reduce((acc, val) => acc.concat(val), []);
}
