// @rx-angular/state/esm2022/lib/rx-state.mjs
import { assertInInjectionContext, DestroyRef, inject } from '@angular/core';
import { RxState as LegacyState } from './rx-state.service';
/**
 * @description
 * Functional way to setup state management with RxState. It's a wrapper around RxState that automatically get
 *   destroyed.
 *
 * @example
 * ```ts
 * import { rxState } from '@rx-angular/state';
 *
 * Component({})
 * export class FooComponent {
 *  readonly state = rxState<{ count: number }>(({ set }) => set({ count: 0 }));
 * }
 * ```
 *
 * @param setupFn
 * @returns RxState instance
 *
 *
 *
 * @docsCategory RxState
 * @docsPage RxState
 *
 */
export function rxState(setupFn) {
    assertInInjectionContext(rxState);
    const legacyState = new LegacyState();
    const destroyRef = inject(DestroyRef);
    destroyRef.onDestroy(() => legacyState.ngOnDestroy());
    const state = {
        get: legacyState.get.bind(legacyState),
        set: legacyState.set.bind(legacyState),
        connect: legacyState.connect.bind(legacyState),
        select: legacyState.select.bind(legacyState),
        signal: legacyState.signal.bind(legacyState),
        computed: legacyState.computed.bind(legacyState),
        computedFrom: legacyState.computedFrom.bind(legacyState),
        $: legacyState.$,
        setAccumulator: legacyState.setAccumulator.bind(legacyState),
    };
    setupFn?.(state);
    return state;
}
