// @rx-angular/state/esm2022/selections/lib/accumulation-observable.mjs
import { BehaviorSubject, EMPTY, merge, queueScheduler, Subject, } from 'rxjs';
import { catchError, distinctUntilChanged, mergeAll, observeOn, publish, publishReplay, scan, tap, withLatestFrom, } from 'rxjs/operators';
export const defaultAccumulator = (st, sl) => {
    return { ...st, ...sl };
};
export function createAccumulationObservable(stateObservables = new Subject(), stateSlices = new Subject(), accumulatorObservable = new BehaviorSubject(defaultAccumulator), scheduler = queueScheduler) {
    const observeStateOn = () => scheduler ? observeOn(scheduler) : (o$) => o$;
    const signal$ = merge(stateObservables.pipe(distinctUntilChanged(), mergeAll(), observeStateOn()), stateSlices.pipe(observeStateOn())).pipe(withLatestFrom(accumulatorObservable.pipe(observeStateOn())), scan((state, [slice, stateAccumulator]) => stateAccumulator(state, slice), {}), tap((newState) => (compositionObservable.state = newState), (error) => console.error(error)), 
    // @Notice We catch the error here as it get lost in between `publish` and `publishReplay`. We return empty to
    catchError((e) => EMPTY), publish());
    const state$ = signal$.pipe(publishReplay(1));
    const compositionObservable = {
        state: {},
        signal$,
        state$,
        nextSlice,
        nextSliceObservable,
        nextAccumulator,
        subscribe,
    };
    // ======
    return compositionObservable;
    // ======
    function nextAccumulator(accumulatorFn) {
        accumulatorObservable.next(accumulatorFn);
    }
    function nextSlice(stateSlice) {
        stateSlices.next(stateSlice);
    }
    function nextSliceObservable(stateObservable) {
        stateObservables.next(stateObservable);
    }
    function subscribe() {
        const sub = compositionObservable.signal$.connect();
        sub.add(compositionObservable.state$.connect());
        sub.add(() => {
            accumulatorObservable.complete();
            stateObservables.complete();
            stateSlices.complete();
        });
        return sub;
    }
}
