// @rx-angular/state/esm2022/actions/lib/proxy.mjs
import { merge, Subject } from 'rxjs';
/**
 * @internal
 * Internal helper to create the proxy object
 * It lives as standalone function because we don't need to carrie it in memory for every ActionHandler instance
 * @param subjects
 * @param transforms
 */
export function actionProxyHandler({ subjectMap, transformsMap, effectMap, errorHandler = null, }) {
    function getEventEmitter(prop) {
        if (!subjectMap[prop]) {
            subjectMap[prop] = new Subject();
        }
        return subjectMap[prop];
    }
    function dispatch(value, prop) {
        subjectMap[prop] = subjectMap[prop] || new Subject();
        try {
            const val = transformsMap && transformsMap[prop]
                ? transformsMap[prop](value)
                : value;
            subjectMap[prop].next(val);
        }
        catch (err) {
            errorHandler?.handleError(err);
        }
    }
    return {
        // shorthand setter for multiple EventEmitter e.g. actions({propA: 1, propB: 2})
        apply(_, __, props) {
            props.forEach((slice) => Object.entries(slice).forEach(([k, v]) => dispatch(v, k)));
        },
        get(_, property) {
            const prop = property;
            // the user wants to get multiple or one single EventEmitter as observable `eventEmitter.prop$`
            if (prop.toString().split('').pop() === '$') {
                // the user wants to get multiple EventEmitter as observable `eventEmitter.$(['prop1', 'prop2'])`
                if (prop.toString().length === 1) {
                    return (props) => merge(...props.map((k) => {
                        return getEventEmitter(k);
                    }));
                }
                // the user wants to get a single EventEmitter as observable `eventEmitter.prop$`
                const propName = prop.toString().slice(0, -1);
                return getEventEmitter(propName);
            }
            // the user wants to get a single EventEmitter and trigger a side effect on event emission
            if (prop.toString().startsWith('on')) {
                // we need to first remove the 'on' from the the prop name
                const slicedPropName = prop.toString().slice(2);
                // now convert the slicedPropName to camelcase
                const propName = (slicedPropName.charAt(0).toLowerCase() +
                    slicedPropName.slice(1));
                return (behaviour, sf) => {
                    const sub = getEventEmitter(propName).pipe(behaviour).subscribe(sf);
                    effectMap[propName] = sub;
                    return () => sub.unsubscribe();
                };
            }
            // the user wants to get a dispatcher function to imperatively dispatch the EventEmitter
            return (args) => {
                dispatch(args, prop);
            };
        },
        set() {
            throw new Error('No setters available. To emit call the property name.');
        },
    };
}
